elf/arch-riscv.cc

```diff
@@ -458,6 +462,49 @@ void InputSection<E>::apply_reloc_alloc(Context<E> &ctx, u8 *base) {
         set_rs1(loc, 4);
       break;
     }
+    case R_RISCV_TLSDESC_HI20: {
+      if (removed_bytes == 0) {
+        u64 desc = sym.get_tlsdesc_addr(ctx);
+        write_utype(loc, desc + A - P);
+      }
+      break;
+    }
+    case R_RISCV_TLSDESC_LOAD_LO12:
+    case R_RISCV_TLSDESC_ADD_LO12: {
+      i64 idx2 = find_paired_reloc();
+      const ElfRel<E> &rel2 = rels[idx2];
+      Symbol<E> &sym2 = *file.symbols[rel2.r_sym];
+      u64 S = sym2.get_addr(ctx);
+      u64 A = rel2.r_addend;
+      u64 P = get_addr() + rel2.r_offset - get_r_delta(idx2);
+      if (sym2.has_tlsdesc(ctx)) {
+        u64 desc = sym2.get_tlsdesc_addr(ctx);
+        write_itype(loc, desc + A - P);
+      } else {
+        if (removed_bytes == 0) {
+          assert(rel.r_type == R_RISCV_TLSDESC_ADD_LO12);
+          *(ul32 *)loc = 0x537; // lui a0,<hi20>
+          write_utype(loc, S + A - ctx.tp_addr);
+        }
+      }
+      break;
+    }
+    case R_RISCV_TLSDESC_CALL: {
+      i64 idx2 = find_paired_reloc();
+      const ElfRel<E> &rel2 = rels[idx2];
+      Symbol<E> &sym2 = *file.symbols[rel2.r_sym];
+      u64 S = sym2.get_addr(ctx);
+      u64 A = rel2.r_addend;
+      if (!sym2.has_tlsdesc(ctx)) {
+        u64 val = S + A - ctx.tp_addr;
+        if (sign_extend(val, 11) == val)
```

Comment:
Not satisfied with how we're computing this twice (here and in shrink_section), so I'm open to suggestions for how to deduplicate this logic.